File: mobile-flutter/lib/features/child/data/child_repository.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';

import '../../../core/api/api_client.dart';
import '../../../models/models.dart';

class ChildRepository {
  ChildRepository({ApiClient? api}) : _api = api ?? ApiClient.instance;
  final ApiClient _api;

  Future<List<ChildProfile>> list([String? userId]) async {
    final data = await _api.get('/api/children') as List;
    return data
        .map((item) {
          final map = Map<String, dynamic>.from(item as Map);
          return ChildProfile.fromMap('${map['id']}', map);
        })
        .toList();
  }

  Future<ChildProfile> create(
    String userId,
    String name,
    int age,
    String gender,
    String note, {
    DevelopmentCategoryKey primaryDifficulty = DevelopmentCategoryKey.other,
    List<LearningGoalKey> learningGoals = const [],
    SupportLevel supportLevel = SupportLevel.medium,
    int dailyDurationMinutes = 5,
    CoLearningMode coLearningMode = CoLearningMode.parentChildTogether,
    String? avatarUrl,
    String? avatarObjectKey,
  }) async {
    final data = await _api.post('/api/children', {
      'displayName': name.trim(),
      'name': name.trim(),
      'age': age,
      'gender': gender,
      'note': note.trim(),
      'primaryDifficulty': primaryDifficulty.apiValue,
      'secondaryDifficulties': <String>[],
      'learningGoals': learningGoals.map((item) => item.apiValue).toList(),
      'supportLevel': supportLevel.apiValue,
      'dailyDurationMinutes': dailyDurationMinutes,
      'coLearningMode': coLearningMode.apiValue,
      'interests': <String>[],
      'accessibilityPreferences': <String, dynamic>{},
      'avatarUrl': avatarUrl ?? '',
      'avatarObjectKey': avatarObjectKey ?? '',
    }) as Map<String, dynamic>;
    return ChildProfile.fromMap('${data['id']}', data);
  }

  Future<void> update(ChildProfile child) async {
    await _api.put('/api/children/${child.id}', child.toMap());
  }

  Future<Map<String, dynamic>> uploadAvatar(
    String childId,
    List<int> fileBytes,
    String fileName,
  ) async {
    final uri = Uri.parse('${_api.baseUrl}/api/children/$childId/avatar');
    final request = http.MultipartRequest('POST', uri);

    final session = _api.session;
    if (session?.accessToken.isNotEmpty == true) {
      request.headers['Authorization'] = 'Bearer ${session!.accessToken}';
    }

    final contentType = fileName.endsWith('.png')
        ? 'image/png'
        : fileName.endsWith('.webp')
            ? 'image/webp'
            : 'image/jpeg';

    final multipartFile = http.MultipartFile.fromBytes(
      'file',
      fileBytes,
      filename: fileName,
      contentType: MediaType.parse(contentType),
    );

    request.files.add(multipartFile);

    final streamedResponse = await request.send();
    final response = await http.Response.fromStream(streamedResponse);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      String errMsg = 'Tải ảnh thất bại: ${response.statusCode}';
      try {
        final bodyJson = jsonDecode(response.body);
        if (bodyJson is Map && bodyJson['message'] != null) {
          errMsg = bodyJson['message'];
        }
      } catch (_) {}
      throw Exception(errMsg);
    }

    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<void> saveCurrentPath(
    String childId,
    String programId,
    String pathId,
  ) async {
    await _api.put('/api/children/$childId/current-path', {
      'programId': programId,
      'pathId': pathId,
    });
  }

  Future<ChildDevelopmentProfile> getDevelopmentProfile(String childId) async {
    final data = await _api.get('/api/children/$childId/development-profile')
        as Map<String, dynamic>;
    return ChildDevelopmentProfile.fromMap(data);
  }

  Future<ChildDevelopmentProfile> updateDevelopmentProfile(
    String childId,
    Map<String, dynamic> payload,
  ) async {
    final data = await _api.put(
      '/api/children/$childId/development-profile',
      payload,
    ) as Map<String, dynamic>;
    return ChildDevelopmentProfile.fromMap(data);
  }
}

File: backend/src/main/java/com/projectha/child/ChildRepository.java
package com.projectha.child;

import com.projectha.common.Db;
import com.projectha.common.NotFoundException;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

@Repository
public class ChildRepository {
    private final JdbcTemplate jdbc;

    public ChildRepository(JdbcTemplate jdbc) {
        this.jdbc = jdbc;
    }

    public List<Map<String, Object>> list(UUID userId) {
        return Db.rows(jdbc.queryForList("SELECT * FROM children WHERE user_id = ? ORDER BY created_at", userId));
    }

    public Map<String, Object> requireOwned(UUID userId, UUID childId) {
        List<Map<String, Object>> rows = jdbc.queryForList(
            "SELECT * FROM children WHERE id = ? AND user_id = ?",
            childId, userId
        );
        return rows.stream().findFirst().map(Db::row).orElseThrow(() -> new NotFoundException("Hồ sơ trẻ không thuộc tài khoản này."));
    }

    public Map<String, Object> create(UUID userId, Map<String, Object> payload) {
        String name = str(payload, "displayName", str(payload, "name", ""));
        return Db.row(jdbc.queryForMap("""
            INSERT INTO children(
              user_id, display_name, name, age, gender, note, primary_difficulty,
              secondary_difficulties, learning_goals, support_level, daily_duration_minutes,
              co_learning_mode, interests, accessibility_preferences, avatar_url, avatar_object_key
            )
            VALUES (?, ?, ?, ?, ?, ?, ?, CAST(? AS jsonb), CAST(? AS jsonb), ?, ?, ?, CAST(? AS jsonb), CAST(? AS jsonb), ?, ?)
            RETURNING *
            """,
            userId,
            name,
            name,
            number(payload.get("age"), 0),
            str(payload, "gender", ""),
            str(payload, "note", ""),
            str(payload, "primaryDifficulty", "OTHER"),
            Db.json(payload.getOrDefault("secondaryDifficulties", List.of())),
            Db.json(payload.getOrDefault("learningGoals", List.of())),
            str(payload, "supportLevel", "MEDIUM"),
            number(payload.get("dailyDurationMinutes"), 5),
            str(payload, "coLearningMode", "PARENT_CHILD_TOGETHER"),
            Db.json(payload.getOrDefault("interests", List.of())),
            Db.json(payload.getOrDefault("accessibilityPreferences", Map.of())),
            str(payload, "avatarUrl", str(payload, "avatar_url", "")),
            str(payload, "avatarObjectKey", str(payload, "avatar_object_key", ""))
        ));
    }

    public Map<String, Object> update(UUID userId, UUID childId, Map<String, Object> payload) {
        requireOwned(userId, childId);
        String name = str(payload, "displayName", str(payload, "name", ""));
        return Db.row(jdbc.queryForMap("""
            UPDATE children SET
              display_name = COALESCE(NULLIF(?, ''), display_name),
              name = COALESCE(NULLIF(?, ''), name),
              age = ?,
              gender = ?,
              note = ?,
              primary_difficulty = ?,
              secondary_difficulties = CAST(? AS jsonb),
              learning_goals = CAST(? AS jsonb),
              support_level = ?,
              daily_duration_minutes = ?,
              co_learning_mode = ?,
              interests = CAST(? AS jsonb),
              accessibility_preferences = CAST(? AS jsonb),
              current_program_id = COALESCE(CAST(NULLIF(?, '') AS uuid), current_program_id),
              current_path_id = COALESCE(CAST(NULLIF(?, '') AS uuid), current_path_id),
              selected_at = COALESCE(CAST(NULLIF(?, '') AS timestamptz), selected_at),
              avatar_url = ?,
              avatar_object_key = ?
            WHERE id = ? AND user_id = ?
            RETURNING *
            """,
            name,
            name,
            number(payload.get("age"), 0),
            str(payload, "gender", ""),
            str(payload, "note", ""),
            str(payload, "primaryDifficulty", "OTHER"),
            Db.json(payload.getOrDefault("secondaryDifficulties", List.of())),
            Db.json(payload.getOrDefault("learningGoals", List.of())),
            str(payload, "supportLevel", "MEDIUM"),
            number(payload.get("dailyDurationMinutes"), 5),
            str(payload, "coLearningMode", "PARENT_CHILD_TOGETHER"),
            Db.json(payload.getOrDefault("interests", List.of())),
            Db.json(payload.getOrDefault("accessibilityPreferences", Map.of())),
            str(payload, "currentProgramId", ""),
            str(payload, "currentPathId", ""),
            str(payload, "selectedAt", ""),
            str(payload, "avatarUrl", str(payload, "avatar_url", "")),
            str(payload, "avatarObjectKey", str(payload, "avatar_object_key", "")),
            childId,
            userId
        ));
    }

    public Map<String, Object> saveCurrentPath(UUID userId, UUID childId, UUID programId, UUID pathId) {
        requireOwned(userId, childId);
        return Db.row(jdbc.queryForMap("""
            UPDATE children
            SET current_program_id = ?, current_path_id = ?, selected_at = now()
            WHERE id = ? AND user_id = ?
            RETURNING *
            """, programId, pathId, childId, userId));
    }

    public void updateAvatar(UUID userId, UUID childId, String avatarUrl, String avatarObjectKey) {
        jdbc.update(
            "UPDATE children SET avatar_url = ?, avatar_object_key = ? WHERE id = ? AND user_id = ?",
            avatarUrl, avatarObjectKey, childId, userId
        );
    }

    private static String str(Map<String, Object> map, String key, String fallback) {
        Object value = map.get(key);
        return value == null ? fallback : String.valueOf(value).trim();
    }

    private static int number(Object value, int fallback) {
        if (value instanceof Number n) return n.intValue();
        try {
            return value == null ? fallback : Integer.parseInt(String.valueOf(value));
        } catch (Exception e) {
            return fallback;
        }
    }
}
